src/transfer.jl

```diff
@@ -1,120 +1,210 @@
+function spherical_transfer(α, S, τ_ref, α_ref, radii, μ_surface_grid)
+    R, r0 = radii[1], radii[end] #lower bound of atmosphere
+
+    #precompute for use in transfer integral
+    log_τ_ref = log.(τ_ref) 
+    τ_ref_α_ref = τ_ref ./ α_ref
+
+    #general element type with which to preallocate arrays
+    el_type = typeof(promote(radii[1], α[1], S[1], μ_surface_grid[1])[1])
+
+    #preallocations
+    #geometric path-length correction, and other wavelength-indenpendent integrand factors
+    integrand_factor = Matrix{el_type}(undef, length(μ_surface_grid), size(α, 1))
+    #the index of the lowest atmospheric layer pierced by each ray
+    lowest_layer_indices = Vector{Int}(undef, length(μ_surface_grid))
     for (μ_ind, μ_surface) in enumerate(μ_surface_grid)
         # impact parameter of ray
         b = R * sqrt(1 - μ_surface^2)
         
-        #calculate the index of the lowest layer the ray passes through
-        i = argmin(abs.(radii .- b))
+        #doing this with `findfirst` is messier at first and last index
+        i = argmin(abs.(radii .- b)) 
         if radii[i] < b
             i -= 1
         end
-        if i == 1 || i == 0
-            I[:, μ_ind] .= 0
-            continue
-        end #can eliminate for i == 1?
-        
-        #ds is the path length through each layer
-        ds = -diff(@. sqrt([R ; radii[1:i]]^2 - b^2)) 
-        dτ = α[1:i, :] .* ds
-        τ = cumsum(dτ, dims=1)
-        
-        for j in 1:size(τ, 2) #iterate over wavelength
-            I[j, μ_ind] = transfer_integral(τ[:, j], S[1:i, j]; plane_parallel=false)
+        lowest_layer_indices[μ_ind] = i
+
+        integrand_factor[μ_ind, 1:i] = @. radii[1:i] ./ sqrt(radii[1:i]^2 - b^2) * τ_ref_α_ref[1:i]
+    end
+
+    #preallocations
+    I = Matrix{el_type}(undef, size(α, 2), length(μ_surface_grid)) #the surface intensity
+    integrand = Vector{el_type}(undef, size(α, 1))                 #integrand of τ integral 
+    τ_λ = Vector{el_type}(undef, size(α, 1))                       #optical depth at a particular λ
+    for λ_ind in 1:size(α, 2), μ_ind in 1:length(μ_surface_grid)
+        i = lowest_layer_indices[μ_ind]
+        for k in 1:i #I can't figure out how to write this as a fast one-liner
+            integrand[k] = α[k, λ_ind] * integrand_factor[μ_ind, k]
         end
-        if b > r0
-            #if the ray never leaves the model atmosphere, include 
-            #the contribution from the other side of the star
-            τ_prime = τ[end:end, :] .+ [cumsum(reverse(dτ[2:end, :], dims=1), dims=1) 
-                                        ; zeros(size(τ[end:end, :]))]
-            for j in 1:size(τ, 2) #Iterate over wavelength
-                I[j, μ_ind] += transfer_integral(τ_prime[:, j], S[1:i, j]; plane_parallel=false)
-            end
-        else #otherwise assume I=S at atmosphere lower boundary
-            I[:, μ_ind] += @. exp(-τ[end, :]) * S[end, :]
+        cumulative_trapezoid_rule!(τ_λ, log_τ_ref, integrand, i)
```

Comment:
Maybe consider adding a comment like the following?
```julia
        cumulative_trapezoid_rule!(τ_λ, log_τ_ref, integrand, i) # compute τ_λ
```

I totally get it if you choose not to do so (since it is julia's convention for functions ending in `!` to mutate the first argument)